src/SubscriptionHooks.php

```diff
@@ -13,240 +13,68 @@
-	 * @param object &$article Article
-	 * @param object &$output Output
-	 * @param object &$user User
-	 * @param object $request WebRequest
-	 * @param object $mediaWiki Mediawiki
-	 *
-	 * @return bool True
+	 * Handle adding premium flair
+	 * @inheritDoc
 	 */
-	public static function onBeforeInitialize( &$title, &$article, &$output, &$user, $request, $mediaWiki ) {
-		global $wgFullHTTPSExperiment;
-
-		if ( defined( 'MW_API' ) && MW_API === true ) {
+	public function onHtmlPageLinkRendererEnd( $linkRenderer, $target, $isKnown, &$text, &$attribs, &$ret ) {
+		$defaultText = trim( strip_tags( HtmlArmor::getHtml( $text ) ) );
+		if ( empty( $target ) ||
+			empty( $target->getText() ) ||
+			$target->getNamespace() !== NS_USER ||
+			!str_starts_with( $defaultText, $target->getText() ) ) {
 			return true;
 		}
 
-		list( $specialPage, ) = MediaWikiServices::getInstance()->getSpecialPageFactory()
-			->resolveAlias( $title->getDBkey() );
-
-		$secureSpecialPages = [ 'Userlogin', 'Preferences' ];
-
-		MediaWikiServices::getInstance()->getHookContainer()->run( 'SecureSpecialPages', [ &$secureSpecialPages ] );
-		if ( $wgFullHTTPSExperiment || ( !empty( $user ) && $user->getId() && in_array( $specialPage, $secureSpecialPages ) ) ) {
-			if ( $request->getProtocol() !== 'https' ) {
-				$redirect = $request->getFullRequestURL();
-				if ( strpos( $request->getFullRequestURL(), 'http://' ) === 0 ) {
-					$redirect = substr_replace( $redirect, 'https://', 0, 7 );
+		$flairClasses = false;
+		if ( array_key_exists( $target->getText(), self::$linkCache ) ) {
+			$flairClasses = self::$linkCache[$target->getText()];
+		} else {
+			$userIdentity = $this->userIdentityLookup->getUserIdentityByName( $target->getText() );
+
+			if ( $userIdentity && $userIdentity->isRegistered() ) {
+				$flairClasses = $this->subscription->getFlairClasses( $userIdentity->getId() );
+				if ( empty( $flairClasses ) ) {
+					// Enforce sanity.
+					$flairClasses = false;
 				}
-				$output->enableClientCache( false );
-				$output->redirect( $redirect, ( $request->wasPosted() ? '307' : '302' ) );
 			}
-
-			return true;
 		}
 
-		if ( ( empty( $user ) || $user->isAnon() ) && !in_array( $specialPage, $secureSpecialPages ) && $request->getProtocol() !== 'http' ) {
-			$response = $request->response();
-			$config = MediaWikiServices::getInstance()->getConfigFactory()->makeConfig( 'main' );
-			$response->clearCookie(
-				'forceHTTPS',
-				[
-					'prefix' => '',
-					'secure' => false,
-					'path' => $config->get( 'CookiePath' ),
-					'domain' => $config->get( 'CookieDomain' ),
-					'httpOnly' => $config->get( 'CookieHttpOnly' )
-				]
-			);
-
-			$redirect = $request->getFullRequestURL();
-			if ( strpos( $request->getFullRequestURL(), 'https://' ) === 0 ) {
-				$redirect = substr_replace( $redirect, 'http://', 0, 8 );
-			}
-			$output->enableClientCache( false );
-			$output->redirect( $redirect, ( $request->wasPosted() ? '307' : '302' ) );
```

Comment:
This hook redirect users to HTTP. This functionality was disabled by flag `"FullHTTPSExperiment": true`